File: utils-debug/src/main/java/nl/ngti/androscope/service/AndroscopeServerHelper.java
package nl.ngti.androscope.service;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.IOException;

import nl.ngti.androscope.server.AndroscopeHttpServer;

final class AndroscopeServerHelper {

    private final AndroscopeHttpServer mServer;
    private final Callback mCallback;

    private AndroscopeServerHelper(@NonNull AndroscopeHttpServer server,
                                   @NonNull Callback callback) {

        mServer = server;
        mCallback = callback;
    }

    @NonNull
    static AndroscopeServerHelper newInstance(@NonNull Context context,
                                              @NonNull Callback callback) {

        final AndroscopeHttpServer server = AndroscopeHttpServer.newInstance(context);
        return new AndroscopeServerHelper(server, callback);
    }

    void start() {
        if (mServer.isAlive()) {
            mCallback.onAlreadyRunning(mServer);
            return;
        }
        mCallback.onStarting();
        try {
            mServer.start();
            mCallback.onStarted(mServer);
        } catch (IOException e) {
            mCallback.onError(e);
        }
    }

    void stop() {
        mServer.stop();

        mCallback.onStopped();
    }

    interface Callback {

        void onStarting();

        void onStarted(AndroscopeHttpServer server);

        void onAlreadyRunning(AndroscopeHttpServer server);

        void onError(IOException e);

        void onStopped();
    }
}

File: lib-androscope/src/main/java/nl/ngti/androscope/service/AndroscopeServerHelper.kt
package nl.ngti.androscope.service

import android.content.Context
import android.os.Build
import android.system.ErrnoException
import android.system.OsConstants
import nl.ngti.androscope.common.log
import nl.ngti.androscope.server.AndroscopeHttpServer
import nl.ngti.androscope.utils.AndroscopeMetadata
import java.io.IOException
import java.net.BindException

internal class AndroscopeServerHelper constructor(
        private val context: Context,
        private val callback: Callback
) {
    private var server: AndroscopeHttpServer? = null

    @Synchronized
    fun start() {
        server?.run {
            if (isAlive) {
                callback.onAlreadyRunning(this)
                return
            }
        }

        callback.onStarting()

        val metadata = AndroscopeMetadata.fromContext(context)
        val portRange = metadata.httpPortRange

        server = attemptToStartServer(portRange, metadata)
    }

    @Synchronized
    fun stop() {
        server?.stop()
        callback.onStopped()
    }

    private fun attemptToStartServer(
            portRange: IntRange,
            metadata: AndroscopeMetadata
    ): AndroscopeHttpServer? {

        var lastException: IOException? = null

        portRange.forEach { port ->
            log { "Trying port $port" }
            val server = AndroscopeHttpServer(context, port, metadata)

            try {
                server.start()
                callback.onStarted(server)
                return server
            } catch (e: IOException) {
                server.stop()

                lastException = e
                if (!shouldRetry(e)) {
                    callback.onError(e)
                    return null
                }
            }
        }
        callback.onError(
                IOException(
                        "Failed to start Androscope within port range $portRange.",
                        lastException)
        )
        return null
    }

    private fun shouldRetry(e: Throwable): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP &&
                e is ErrnoException) {
            return e.errno == OsConstants.EADDRINUSE
        }
        if (e is BindException) {
            return e.cause?.let { shouldRetry(it) } ?: false
        }
        return false
    }

    internal interface Callback {
        fun onStarting()
        fun onStarted(server: AndroscopeHttpServer)
        fun onAlreadyRunning(server: AndroscopeHttpServer)
        fun onError(e: IOException)
        fun onStopped()
    }
}
